### ISSUE TITLE
Add defaults for entry and output in WebpackOptionsDefaulter.

### ISSUE THREAD

**Issue by TheLarkInn** [webpack/webpack#2954]
**I'm submitting a feature request**

**Webpack version:**
2.x

**Please tell us about your environment:**
OSX 10.x / Linux / Windows 10

**Current behavior:**
You must specify at minimum an entry property and an output property. 

**Expected/desired behavior:**
Do we really need to? Or could these be generic properties. I imagine the following by default: 

(THIS IS JUST AN EXAMPLE, not what it really needs to be).

``` javascript
entry: {
  main: './src/index.js'
},
output: {
  filename: '[name].bundle.js',
  path: 'dist'
}
```
- **What is the motivation / use case for changing the behavior?**
  tl;dr: @gaearon suggested this to me and I agree that we could make a sensible default for these two values. In the end do things like input and output always get changed, most likely, yes. But there's also no reason to not have a default. 
- **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ] 
- **Language:** [all | TypeScript X.X | ES6/7 | ES5 | Dart | ...]

**Comment by gdi2290** [comment]
Yes please. Perhaps check `src/index` then `index`

**Comment by theverything** [comment]
For the entry you could check `package.json` for the `main` field. For output I wouldn't assume anything about the folder structure of the project and maybe just output the bundle to the `cwd`.

**Comment by gaearon** [comment]
I would warn against trying to support 5 different popular conventions. People are confused by Webpack precisely because its inputs are too relaxed and break subtly in unexpected ways. Pick one convention and run with it. It's just a default after all.

As for looking for "main" field this could be confusing because "main" should usually point to compiled output, not the source.

**Comment by NekR** [comment]
I don't see how it's useful. Maybe it's good to have some default output (`[name].bundle.js`) and then output it to the same folder of input file, but input should be explicit. After all, no normally work with webpack, you still need to provide more accurate configuration.

> What is the motivation / use case for changing the behavior?
tl;dr $@gaearon

What is the real reason?

**Comment by danekszy** [comment]
@NekR Isn't it just an attempt to lower the entry barrier?
Most users will be knowledgable enough to change their config, but beginners will have it easier to just test the waters...
Additional benefit is that even as a _"poweruser"_ you'll be able to kickstart bundling, whenever you won't need any bells & whistles.
👍 from me... 😄

**Comment by NekR** [comment]
I don't see how guide:
> Put your entry into `src/index.js` and then run `webpack`

is easier than:

> Run `webpack <entry>` where `<entry>` is your file, e.g. `index.js` or `src/index.js`

To me this seems more like one more way to things, which is most complainable thing about wepback. This doesn't simply things, just adds one more way.
______________________________
On the other side, while it's also not hard to do `webpack input.js output.js`, it feels more natural to have default output option. The reasons is that most of the compilers/tools out there (not on JS related) have default output option which is inferred from input one. However, it's very-very rare for such tools to have default input option.

**Comment by ericclemmons** [comment]
Shouldn't the entry be based on `target`?  Web is the default, but node is another common build target where defaulting to `src/index.js` wouldn't make sense.

**Comment by gaearon** [comment]
To clarify: I don't really care about `entry` but I think default sane `output` is essential.

**Comment by TheLarkInn** [comment]
And @nekr I'll update the issue motivation. Was trying to elude to Dan bringing this up to me as am idea, and that I think it's pragmatic at least.

**Comment by TheLarkInn** [comment]
@NekR updated motivations section for more clarification!

**Comment by NekR** [comment]
@gaearon I'm with having default for `ouput`. I suggest do that and leave `entry` alone. 
@TheLarkInn thanks for that. By _reason_ I meant real reason which you found out in the wild. I understood that Dan suggested it from first version.

 Adding feature just because you can doesn't sound like a pragmatic decision. What gives default entry apart from limitation of where to put entry file? Do you need it for tutorials, demos?

**Comment by kentcdodds** [comment]
If we're voting, I vote for this:

```
entry: './src/index.js'
output: {
  filename: 'bundle.js',
  path: 'dist'
}
```

Most apps wont need multiple entries.

**Comment by pago** [comment]
As might be obvious from my PR for this, I'm all for having defaults.

The current default value for the output filename is `[name].js` and the path is set to the current working directory. Allowing default values is simply forbidden within the CLI.
I'd recommend to go with a name that includes the `[name]` pattern since a configuration that only contains an `entry` part would then be valid.

With the suggestion from @kentcdodds, the following config would be valid:

```
entry: './src/index.js'
```

while this one would not be valid:

```
entry: {
   main: './src/index.js',
   aside: './src/aside.js'
}
```

Personally, I prefer `dist/[name].js` over `dist/[name].bundle.js` as the `bundle` part only adds noise. Its already obvious that we're dealing with the build result from the `dist` directory name, why would we need another pointer to that by using the `.bundle` infix?

I don't see any harm in choosing `src/index.js` as the default entry point. But that might be because that's usually the entry point I use.

**Comment by TheLarkInn** [comment]
@pago I think I agree. One of the things that has been on my mind lately is the confusion of people not understanding the difference between a `chunk` and a `bundle`, or really the misconception that they are practically the same thing. I think the `.bundle` infix was a common pattern used by the community as webpacks popularity developed.

**Comment by TheLarkInn** [comment]
Alright we should document all the points that need to be address through a feature like this. It sounds crazy, but many of you have already brought up great points that need to be addressed. 

## Adding Entry and Output Defaults
* Identify affected areas by this change: 
  * For example, a bit of work on internal tests would have to be updated to cater to assigned values of `entry` and `output`.
  * This addition to defaults may not be a breaking change, but they certainly could lead to unexpected results, or confusion if not imperatively simple. 
* Should or do these values change based upon `target`. This is an attempt at a pragmatic configuration default, so do we only have one case for when target is also being defaulted? 
* What should `entry` be and why? `./index.js` feels right? Worth discussion.
* What should `output` be. `[name].bundle.js` has been thrown around? `[name].chunk.js`? `[name].js`?

Additions should be made, I can update this comment and place its content into the issue itself.

**Comment by Kovensky** [comment]
Also, why should the entry be "./src/index.js", and not "./src" or "./src/"?

**Comment by TheLarkInn** [comment]
Agreed, it could just resolve to index.

**Comment by andreypopp** [comment]
I'd go **against** a default value for `entry`:

* For some reason executing just `webpack` feels more magical than `webpack ./src/index.js`. Maybe this is me but this looks too implicit and thus confusing.
* It doesn't seem to be the widely used convention to have `./src/index.js` as an entry.
* The fact that we need to think if we need different defaults for `entry` depending on `target` feels wrong. That does suggest that there's no default.

I'm all for having a default value for `output` but I'd rather make it spur all build artefacts into a directory, for example `build/`. That way it is easy to get rid of them — just `rm -r build/` — I think this is important.

**Comment by ericclemmons** [comment]
@andreypopp Has persuaded me otherwise.  There's largely _not_ a convention (and it looks like CRA is hoping to create it), but `./src/index.js` is fine for a published module, not for a Webpack project that can easily grow to require multiple entry-points.  **And by that time, the names of those entries are project-specific and not based on `target`**.

The one thing I've noticed with isomorphic apps with Webpack is that it's wise to have resources output in subfolders of `build` (e.g.`build/client`, `build/server`, etc.). Because several projects serve public assets out of the build folder, **you want to ensure you don't serve potentially sensitive, server-side source code**.

In hindsight, `webpack [entry]` isn't asking too much.

**Comment by kentcdodds** [comment]
I think that we should probably consider the "default use case" for this. What are _most_ people using webpack for? That's what we should be considering when it comes to good defaults.

I would argue that the main use-case is for client-only applications. So we shouldn't be considering different targets or universal apps at all when deciding on good default.s

**Comment by ericclemmons** [comment]
CRA, [`react-boilerplate`](https://github.com/mxstbr/react-boilerplate) and countless other boilerplates differ in both entry-point name and which folder it's in.

I doubt there's a consensus without us picking a side & banking on it, in which case we should pick a defensible answer to _both_ `entry` and `output` as defaults.  (Which is attainable, but I'll gladly bow out for others to decide for me :D)

**Comment by dhruvdutt** [comment]
@TheLarkInn We can close this now since webpack 4 covers it. :fire: :rocket:
